### PR TITLE
Reenable TCL/Tk by disabling SQLLite dependency

### PR DESCRIPTION
**Pull request by highperformancecoder**
TCL was originally disabled because of a compiler build error in the SQL lite component.
SQLLite functionality is not an important feature of TCL, and can be disabled via a configuration variable.
TCL/Tk is an important component technology for projects of mine that use MXE, so I will be maintaining TCL/Tk indefinitely, at least within my branch.

Only commits 848c83ea3fb99101196a8c14409bc5bcef2ee5ee and dbe181a01dc796a9ca19b6c0611a2130d86a4f4b are needed for this pull request. I can't seem to specify which commits to include in a pull request via Github's pull request create form.

5dbde9debfb897a8e40352531eb109018e30e0c4 is for a compile problem I found in qdbm.

### CODE REVIEW

File: src/tk.mk [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    $(SED) -n 's,.*/\([0-9][^"]*\)/".*,\1,p' | \
+    head -1
+endef
+
+define $(PKG)_BUILD
+    cd '$(1)/win' && ./configure \
+        $(MXE_CONFIGURE_OPTS) \
+        --enable-threads \
+	$(if $(findstring 64,$(TARGET)), --enable-64bit) \
+	CFLAGS=-D__MINGW_EXCPT_DEFINE_PSDK
```
Comment (TimothyGu) [inline src/tk.mk]: Indentation is off.

File: src/tk.mk [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    $(MAKE) -C '$(1)/win' install bin_PROGRAMS= sbin_PROGRAMS= noinst_PROGRAMS=
+endef
+
+# tcl doesn't compile on i686-pc-mingw32. See
+# https://github.com/mxe/mxe/issues/508
+$(PKG)_BUILD_i686-pc-mingw32 :=
```
Comment (TimothyGu) [inline src/tk.mk]: You can now remove lines 27-30 since we removed support for i686-pc-mingw32.

File: src/tcl.mk [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        $(MXE_CONFIGURE_OPTS) \
+        --enable-threads --without-sqlite\
+        $(if $(findstring 64,$(TARGET)), --enable-64bit) \
+        CFLAGS='-D__MINGW_EXCPT_DEFINE_PSDK'
+    $(MAKE) -C '$(1)/win' install install-private-headers $(MXE_DISABLE_PROGRAMS)
+endef
+
+# tcl doesn't compile on i686-pc-mingw32. See
+# https://github.com/mxe/mxe/issues/508
+#$(PKG)_BUILD_i686-pc-mingw32 :=
```
Comment (TimothyGu) [inline src/tcl.mk]: You can now remove lines 28-31 since we removed support for i686-pc-mingw32.

**Comment by highperformancecoder** [comment]
Hi Tim, I found quite a number of packages that no longer build, for a variety of reasons, including source code having moved repositories. I fixed a few of them, hopefully correctly, but I have decided now to just build those packages I directly need, and fix those if they're broken, and just not worry about the rest, as I don't have enough time for this!

But kudos anyway, as MXE really helps my work.

**Comment by tonytheodore** [comment]
I couldn't get tcl to build correctly, the static builds produce dlls and the shared build fails.

If you get them working again, please open a new request (hopefully with only the relevant commits).

**Comment by highperformancecoder** [comment]
Hi Tony, could you please explain a bit more what you mean? The static build works - ie creates the static .a libraries. It well also create dynamic libs, but is that a problem? What do you mean by shared build? I thought MXE only build static libraries.

**Comment by highperformancecoder** [comment]
BTW - how do you stop github from adding successor commits to a pull request?

**Comment by kolewu** [comment]
Please read about pull requests: http://lmgtfy.com/?q=what+is+a+pull+request%3F

With a pull request you ask to merge the contents of a whole branch (that is all commits from the pull requests HEAD down to the point where the two branches met) into another one. And working on a pull request means commiting on this branch. So stopping github from adding commits to a pull request is not _the right thing_.

**Comment by highperformancecoder** [comment]
Thanks Uwe. Your first URL is effectively just "Google pull requests", which is a little supercilious :). I had already read the first hit (Github's description), and for good measure I read the second hit (from OSSWatch). Neither page mentions that pull requests work like that. Such is the state of OSS documentation :).

Do you have a link describing the right way to go about doing a pull request, since creating them from my repository in the manner described by the above pages is apparently the wrong thing to do.

Should I, for example, create a staging branch that I selectively merge changesets from, and interpose that in between my repository and the mxe master repository? I think I read somewhere that it is possible to reattach a branch (or is it a whole repository) to a different upstream branch.

**Comment by highperformancecoder** [comment]
Also is there any answer to my question about what Tony means by a shared build in MXE?

**Comment by highperformancecoder** [comment]
I found these suggestions on Stack Overflow:
Couple tips on pull-requests:

 *   create a branch: isolate your modifications in a branch. Don't create a pull request from master, where you could be tempted to accumulate and mix several modifications at once.
 *   rebase that branch: even if you already did a pull request from that branch, rebasing it on top of origin/master (making sure your patch is still working) will update the pull request automagically (no need to click on anything)
 *   update that branch: if your pull request is rejected, you simply can add new commits, and/or redo your history completely: it will activate your existing pull request again.
*    "focus" that branch: i.e., make its topic "tight", don't modify thousands of class and the all app, only add or fix a well-defined feature, keeping the changes small.
*    delete that branch: once accepted, you can safely delete that branch on your fork (and git remote prune origin). The GitHub GUI will propose for you to delete your branch in your pull-request page.

I might try again, but in a branch, so that I can dripfeed tge changesets,

**Comment by kolewu** [comment]
Sorry, the lmgtfy-link was not meant supercilious.  But your post (and pull request) had not looked like you have done your homework.

For me the oss-watch link is the third, but it has the whole story only being a bit reticent about the further work on the pull request (or merge request like other systems call it): _Report any problems to the contributor and request a resubmission_

The 6th link has a nice explanation on this: https://www.thinkful.com/learn/github-pull-request-tutorial/#Rebasing

For my last three pull requests for mxe I have made my changes while doing a full compile and afterwords split them up on three distinct branches, that I then rebased on master.  For working away with the sum of my own changes, I created a working branch and merged all the features branches there.  This is super easy with git, especially if the branches do not overlap.

**Comment by starius** [comment]
@highperformancecoder, the main thing I learned about git, github and pull requests is that committing to `master` branch of my fork is not convenient, because in that case starting a new development needs either waiting until a pull request is merged (which can be long-long time) or forking to another github account. The latter doesn't look good and still limits the number of mutual pull requests by the number of github accounts...

Then I read this short and easy [instruction](https://github.com/Olivine-Labs/busted/blob/master/CONTRIBUTING.md#using-git-appropriately). My main error was ignoring git branches! Now I create a branch for every independent development and it works like a charm.

**Comment by highperformancecoder** [comment]
Thanks guys - I'm still pretty new to git and github (<1 year), so this really is an education opportunity.
In the case of MXE, I will need to have all my changes going into my master, as I have dependent projects that rely on this, but I'll try to split out the changes into separate branches for pull requests to the upstream.

**Comment by starius** [comment]
> I'll try to split out the changes into separate branches for pull requests to the upstream

It is a good idea. I use patch files for that:
```
$ git checkout master
$ git format-patch HEAD~10
```
It creates a patch file per a commit.

Now create a branch based on a commit from upstream.

```
$ git remote add upstream git@github.com:mxe/mxe.git
$ git fetch upstream
$ git checkout -b branch-for-upstream upstream/master
```

Add two commits from patches to that branch and push it.

```
$ git am 0002-msg.patch 0005-foo-bar.patch
$ git push
```

I hope it helped.

**Comment by tonytheodore** [comment]
@highperformancecoder 
> Also is there any answer to my question about what Tony means by a shared build in MXE?

I've forgotten what the issues were, but I'll test it again in #941 and let you know. MXE has had shared builds for some time and we like to keep them separated from the static builds.

To create #941 I just did a `fetch` and `cherry-pick` of the two commits in a new branch as there were too many merge conflicts.

> In the case of MXE, I will need to have all my changes going into my master

That will make it harder to keep updated with MXE (i.e. shared builds and removal of i686-pc-mingw32) and also make it harder to contribute changes. It's worth keeping updated as we will soon have a new way to handle such things via plugins #935 that should remove most of the merge/rebase issues if you want to maintain separate packages or overrides.

@kolewu wrote:
> This is super easy with git, especially if the branches do not overlap.

Any tips on how to handle features that do overlap or rely on previous requests (like #935)?

**Comment by kolewu** [comment]
In theory even repeating overlapping merges with conflicts should be solvable easy in git with [rerere](http://git-scm.com/2010/03/08/rerere.html). But I don't have any hands-on experience with this.

**Comment by highperformancecoder** [comment]
Still trying to figure out how to create pull requests. In my first attempt, I tried creating a branch to create the pull request in, but this didn't fly, because the master branch has multiple changes with respect to the upstream repository. So then I tried forming the upstream mxe repository again  - which github won't do. So then I used a technique adrian short to duplicate forked repositories (https://adrianshort.org/create-multiple-forks-of-a-github-repo), which works to an extent, although now my pull requests go to the master branch of the origin repository, not the configured upstream repository.

Any hints? At this stage, all I can think of doing is to create a completely separate github user account, just to hold a fork of mxe, and the add myself as a collaborator so I can create pull requests on it. But that seems insanely hard - there must be an easier way.

**Comment by highperformancecoder** [comment]
OK, I did the following, which appears to have worked.
1. Renamed the repository with my clean pull requests to mxe-pull-requests2
2. Created a new repository, mxe-master and populated it with the contents of the old mxe fork
3. Deleted the original forked repository
4. Created a new fork of mxe/mxe
5. Renamed the forked mxe repository to mxe-pull-requests
6. Created a new branch to replicate the mxe-pull-requests2 structure.
7. pushed the new changes to mxe-pull-requests
8. created the pull requests against mxe/mxe
9. renamed mxe-master to mxe

**Comment by kolewu** [comment]
Would have been a lot easier to

1. just create a branch at the top of the forked master
2. reset master to origin/master.
3. force push

That's it. You have a branch with all your changes and a clean master that can follow origin.

Now the next steps are:
* separating your changes into features branches
* make a (private) working branch with a merge of all your feature branches
* create feature requests for every feature branch

When there is an update on master (maybe with one of your feature branches merged):
* rebase all remaining feature branches
* reset --hard your working branch an fast-forward to origin/master
* merge the remaining feature branches

If you are unsure with any of this steps, just make a plain copy of the repository (directory) and try there.  As long as you don't perform a garbage collection you can undo all failures with the help of the reflog.

**Pull request closed without merging**